src/plugins/presentation_util/public/components/input_controls/control_types/options_list/options_list_component.tsx

```diff
@@ -0,0 +1,171 @@
+
+import './options_list.scss';
+
+const toggleAvailableOptions = (
+  indices: number[],
+  availableOptions: EuiSelectableOption[],
+  enabled: boolean
+) => {
+  const newAvailableOptions = [...availableOptions];
+  indices.forEach((index) => (newAvailableOptions[index].checked = enabled ? 'on' : undefined));
+  return newAvailableOptions;
+};
+
+interface OptionsListProps {
+  input: OptionsListEmbeddableInput;
+  fetchData: OptionsListDataFetcher;
+}
+
+export const OptionsListInner = ({ input, fetchData }: OptionsListProps) => {
+  const [availableOptions, setAvailableOptions] = useState<EuiSelectableOption[]>([]);
+  const selectedOptions = useRef<Set<string>>(new Set<string>());
+
+  // raw search string is stored here so it is remembered when popover is closed.
+  const [searchString, setSearchString] = useState<string>('');
+  const [debouncedSearchString, setDebouncedSearchString] = useState<string>();
+
+  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
+  const [loading, setIsLoading] = useState(false);
+
+  const typeaheadSubject = useMemo(() => new Subject<string>(), []);
+
+  useMount(() => {
+    typeaheadSubject
+      .pipe(
+        tap((rawSearchText) => setSearchString(rawSearchText)),
+        debounceTime(100)
+      )
+      .subscribe((search) => setDebouncedSearchString(search));
+    // default selections can be applied here...
+  });
+
+  const { indexPattern, timeRange, filters, field, query } = input;
+  useEffect(() => {
+    let canceled = false;
+    setIsLoading(true);
+    fetchData({
+      search: debouncedSearchString,
+      indexPattern,
+      timeRange,
+      filters,
+      field,
+      query,
+    }).then((newOptions) => {
+      if (canceled) return;
+      setIsLoading(false);
+      // We now have new 'availableOptions', we need to ensure the previously selected options are still selected.
+      const enabledIndices: number[] = [];
+      selectedOptions.current?.forEach((selectedOption) => {
+        const optionIndex = newOptions.findIndex(
+          (availableOption) => availableOption.label === selectedOption
+        );
+        if (optionIndex >= 0) enabledIndices.push(optionIndex);
+      });
+      newOptions = toggleAvailableOptions(enabledIndices, newOptions, true);
+      setAvailableOptions(newOptions);
+    });
+    return () => {
+      canceled = true;
+    };
+  }, [indexPattern, timeRange, filters, field, query, debouncedSearchString, fetchData]);
+
+  const updateItem = useCallback(
+    (index: number) => {
+      const item = availableOptions?.[index];
+      if (!item) return;
+
+      const toggleOff = availableOptions[index].checked === 'on';
+
+      const newAvailableOptions = toggleAvailableOptions([index], availableOptions, !toggleOff);
+      setAvailableOptions(newAvailableOptions);
+
+      if (toggleOff) selectedOptions.current.delete(item.label);
+      else selectedOptions.current.add(item.label);
```

Comment:
nit: the two line if-else here is a little hard to read but that might just be me

Reply:
Yeah, some braces here might help readability!